include/llbuild/Basic/FileSystem.h

```diff
@@ -153,6 +163,71 @@ class DeviceAgnosticFileSystem : public FileSystem {
   }
 };
 
+/// Checksum-only filesystem wrapper
+class ChecksumOnlyFileSystem : public FileSystem {
+private:
+  std::unique_ptr<FileSystem> impl;
+
+public:
+  explicit ChecksumOnlyFileSystem(std::unique_ptr<FileSystem> fs)
+    : impl(std::move(fs))
+  {
+  }
+
+  ChecksumOnlyFileSystem(const FileSystem&) LLBUILD_DELETED_FUNCTION;
+  void operator=(const ChecksumOnlyFileSystem&) LLBUILD_DELETED_FUNCTION;
+  DeviceAgnosticFileSystem &operator=(ChecksumOnlyFileSystem&& rhs) LLBUILD_DELETED_FUNCTION;
+
+  static std::unique_ptr<FileSystem> from(std::unique_ptr<FileSystem> fs);
+
+
+  virtual bool
+  createDirectory(const std::string& path) override {
+    return impl->createDirectory(path);
+  }
+
+  virtual bool
+  createDirectories(const std::string& path) override {
+    return impl->createDirectories(path);
+  }
+
+  virtual std::unique_ptr<llvm::MemoryBuffer>
+  getFileContents(const std::string& path) override;
+
+  virtual bool remove(const std::string& path) override {
+    return impl->remove(path);
+  }
+
+  virtual FileChecksum getFileChecksum(const std::string& path) override {
+    return impl->getFileChecksum(path);
+  }
+
+  virtual FileInfo getFileInfo(const std::string& path) override {
+    auto info = impl->getFileInfo(path);
+
+    info.device = 0;
+    info.inode = 0;
+    info.modTime = FileTimestamp();
+
+    info.checksum = impl->getFileChecksum(path);
+
+    return info;
+  }
+
+  virtual FileInfo getLinkInfo(const std::string& path) override {
+    auto info = impl->getLinkInfo(path);
+
+    info.device = 0;
```

Comment:
Right now, symlinks aren't using checksums. Should this avoid clearing the device and node numbers, so that if the underlying implementation is not a device-agnostic filesystem, we don't impose that behavior?

Reply:
I can't see a blocker for symlinks to calculate checksums (but we do need to write tests for them). `ChecksumOnlyFileSystem::getLinkInfo` should look very similar to `ChecksumOnlyFileSystem::getFileInfo` and request for the checksum of the path.

Later on if we wanted to implement a FileInfoCache, we might want to resolve the symlink first to avoid multiple checksumming.

Reply:
I updated `getLinkInfo` so that on Unix we `readlink` and checksum the result of that.

On Windows.. I don't know.. I'm not even sure if `getLinkInfo` will be called at all or not. In absence of the notion of symlinks on that platform, I relied on `fread`ing the path and checksumming the content of that. We can revisit the Windows case later.